File: solutions/_120_triangle/solution.cpp

#include <vector>
using namespace std;

class Solution
{
public:
    int minimumTotal(vector<vector<int>> &triangle)
    {
        int n = triangle.size();
        vector<int> dp(n + 1);
        for (int i = n - 1; i >= 0; i--)
        {
            vector<int> &row = triangle[i];
            for (int j = 0; j < row.size(); j++)
            {
                dp[j] = row[j] + min(dp[j], dp[j + 1]);
            }
        }

        return dp[0];
    }
};

/*
https://leetcode.com/problems/triangle/
*/

File: solutions/_120_triangle/solution.py
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        dp = [0] * (len(triangle) + 1)
        for row in triangle[::-1]:
            for i, num in enumerate(row):
                dp[i] = min(dp[i], dp[i + 1]) + num

        return dp[0]


# https://leetcode.com/problems/triangle/
